### PR TITLE
feat: implement billing service webhook receiver authentication service

### PR DESCRIPTION
**Pull request by LucieFaire**
### Context
The CGW needs to receive webhooks from the billing service and verify their origin. This implements the **"receiver issues the credential it later checks"** model: the CGW mints a long-lived **ES256** (ECDSA P-256) token, provisions it to the billing service, and verifies each incoming `Authorization: Bearer <token>` **statelessly/offline** against its own public key — no JWKS, no callback, no shared secret over the body.

### Verification (runtime)
- **`BillingWebhookAuthGuard`** — thin guard that extracts the bearer token and delegates to `BillingAuthService`.
- **`BillingAuthService.verify`** — two layers: (1) ES256 signature + `iss`/`aud`/`exp` via `IJwtService`; (2) service-token authorization (Zod schema: `roles` contains `SERVICE_ACCESS`, or `data` has `permission_type=SERVICE_ACCESS` + `user_type=SERVICE_USER` + non-empty `service_name`). Failures log at `warn` and return `401`.
- **Endpoint** — placeholder `POST /v1/billing/webhooks` returning `202` (payload handling is a follow-up).
- **Feature-flag gated** — the module is only wired in when `FF_BILLING_WEBHOOK=true`; the app reads its public key lazily. Enabling the flag without a key fails fast at boot. `iss`/`aud` are a single configured identifier.
- The running app **never signs and never calls KMS** — it only verifies, offline, with the public key.

### Provisioning (`yarn generate-token` / `scripts/generate-token.ts`)
Mints the long-lived token. Dual signing mode, same claim shape either way:
- **KMS (production)** — signs via an asymmetric KMS key (`ECC_NIST_P256`); the private key never leaves KMS. Also prints the public-key PEM to configure the verifier.
- **Local PEM (dev/CI)** — signs with `BILLING_WEBHOOK_JWT_PRIVATE_KEY`.
- **Enforced:** when `CGW_ENV=production`, local-PEM signing is refused (KMS required).
- Strict `parseArgs` (unknown flags rejected); `--sub` is charset-validated.

KMS specifics:
- `AwsKmsSignerService` (`src/datasources/kms/`) — `Sign` (ECDSA_SHA_256) + `GetPublicKey`; credentials via IRSA or the default AWS chain; region from `AWS_REGION`.
- `BillingAuthService.mintViaSigner` hand-assembles the JWS around the KMS signature; DER→JOSE conversion via `ecdsa-sig-formatter` (promoted from a transitive to a direct dependency).

### Configuration
| Variable | Used by | Default |
| -- | -- | -- |
| `FF_BILLING_WEBHOOK` | app (enables endpoint) | `false` |
| `BILLING_WEBHOOK_JWT_PUBLIC_KEY` | app (verify) | — |
| `BILLING_WEBHOOK_JWT_ISSUER` | app + CLI (`iss`/`aud`) | `safe-client-gateway` |
| `BILLING_WEBHOOK_JWT_KMS_KEY_ID` | CLI (KMS signing) | — |
| `BILLING_WEBHOOK_JWT_PRIVATE_KEY` | CLI (local signing) | — |

The app needs **no** KMS permissions; `kms:Sign`/`kms:GetPublicKey` go to a dedicated mint principal (CI/bastion), not the app's runtime role. See `src/modules/billing/README.md`.

### Notes
- Webhook **payload** processing is intentionally out of scope (placeholder endpoint).
- The mint CLI is a dev/ops tool; in the deployed image invoke `node dist/scripts/generate-token.js` (not `yarn`) — see the module README for the KMS/IAM provisioning guide.
- Related: PLA-1679 is the opposite direction (CGW → billing outbound credential); keys/tokens are kept separate.

### CODE REVIEW

File: src/modules/billing/routes/guards/billing-webhook-auth.guard.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Protects the billing-service webhook endpoint with a service-to-service JWT.
+ *
+ * Extracts the `Authorization: Bearer <token>` credential (an HTTP concern) and
+ * delegates verification to {@link BillingAuthService}. Only instantiated
+ * when the `billingWebhook` feature is enabled.
+ */
+@Injectable()
+export class BillingWebhookAuthGuard implements CanActivate {
+  constructor(private readonly tokenService: BillingAuthService) {}
```
Comment (moisses89) [inline src/modules/billing/routes/guards/billing-webhook-auth.guard.ts:21]: `BillingWebhookAuthGuard` injects the concrete `BillingAuthService` class, can we consider following the same pattern as `AuthGuard`, which injects an interface `IAuthRepository`?.

File: src/modules/billing/routes/guards/billing-webhook-auth.guard.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    return true;
+  }
+
+  private getBearerToken(request: Request): string | null {
+    const header = request.headers.authorization;
+    if (!header?.startsWith('Bearer ')) {
+      return null;
+    }
+    const token = header.slice('Bearer '.length).trim();
+    return token.length > 0 ? token : null;
```
Comment (moisses89) [inline src/modules/billing/routes/guards/billing-webhook-auth.guard.ts:44]: Just to add coverage, can you add a test for empty key after Bearer?

Comment (LucieFaire) [reply src/modules/billing/routes/guards/billing-webhook-auth.guard.ts:44]: [a3de9e6](https://github.com/safe-global/safe-client-gateway/pull/3217/commits/a3de9e611429f3115331c1433ef9740427bf2c91)

File: src/modules/billing/routes/guards/billing-webhook-auth.guard.ts (hunk 1 above)
Comment (LucieFaire) [reply src/modules/billing/routes/guards/billing-webhook-auth.guard.ts:21]: hey, thanks for pointing out. I feel like calling that class `AuthRepository` was maybe not exactly correct. (an there is an interface primarily cause the convention is: repo + interface).
It only calls JWT methods and doesnt have any data access. If we look at `CapthaGuard`, for instance, it also contains a concrete service. I dont expect us to have diff implementation for billing auth, so I would argue service is a correct choice

File: src/modules/billing/routes/guards/billing-webhook-auth.guard.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      );
+    }
+
+    // Throws UnauthorizedException if the token is invalid or unauthorized.
+    this.tokenService.verify(token);
+    return true;
+  }
+
+  private getBearerToken(request: Request): string | null {
+    const header = request.headers.authorization;
```
Comment (PooyaRaki) [inline src/modules/billing/routes/guards/billing-webhook-auth.guard.ts:39]: Nit: We could make the header name configurable.
```suggestion
   const AUTH_HEADER_NAME = 'Authorization';
    const header = request.headers[AUTH_HEADER_NAME];
```

File: src/modules/billing/routes/guards/billing-webhook-auth.guard.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Protects the billing-service webhook endpoint with a service-to-service JWT.
+ *
+ * Extracts the `Authorization: Bearer <token>` credential (an HTTP concern) and
+ * delegates verification to {@link BillingAuthService}. Only instantiated
+ * when the `billingWebhook` feature is enabled.
+ */
+@Injectable()
+export class BillingWebhookAuthGuard implements CanActivate {
```
Comment (PooyaRaki) [inline src/modules/billing/routes/guards/billing-webhook-auth.guard.ts:20]: We could use or extend the current `AuthGuard` and support passing the auth property as a header value. The code here is mostly duplicated.

File: src/modules/billing/domain/billing-auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  constructor(
+    @Inject(IJwtService) private readonly jwtService: IJwtService,
+    @Inject(IConfigurationService)
+    configurationService: IConfigurationService,
+    @Inject(LoggingService) private readonly loggingService: ILoggingService,
+  ) {
+    this.publicKey = configurationService
+      .getOrThrow<string>('billing.webhook.publicKey')
+      // PEM keys provided via env often arrive with escaped newlines.
+      .replace(/\\n/g, '\n');
```
Comment (PooyaRaki) [inline src/modules/billing/domain/billing-auth.service.ts]: Nit: We can handle it in our validator. We should validate this env variable and transform it if it's valid.

File: src/modules/billing/domain/billing-auth.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+      );
+    });
+
+    it('honours a custom subject and reflects it in service_name', () => {
+      const { privateKey } = generateKeyPair();
+
+      const token = BillingAuthService.mint({
+        privateKey,
+        issuer: ISSUER,
+        expiresInDays: 365,
```
Comment (PooyaRaki) [inline src/modules/billing/domain/billing-auth.service.spec.ts]: We should use a random number

File: src/modules/billing/domain/billing-auth.service.spec.ts [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+          sub: 'billing-service-staging',
+          data: { service_name: 'billing-service-staging' },
+        },
+      );
+    });
+
+    it('derives iat/exp from the issued-at time and expiresInDays', () => {
+      const { privateKey } = generateKeyPair();
+      // Whole seconds so the second-based NumericDate comparison is exact.
+      const now = new Date(1_700_000_000_000);
```
Comment (PooyaRaki) [inline src/modules/billing/domain/billing-auth.service.spec.ts]: We should use `faker`.

File: src/modules/billing/domain/billing-auth.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+import { UnauthorizedException } from '@nestjs/common';
+import type { MockedObject } from 'vitest';
+import { FakeConfigurationService } from '@/config/__tests__/fake.configuration.service';
+import { JWT_ES_ALGORITHM } from '@/datasources/jwt/jwt.constants';
+import { jwtClientFactory } from '@/datasources/jwt/jwt.module';
+import { JwtService } from '@/datasources/jwt/jwt.service';
+import type { ILoggingService } from '@/logging/logging.interface';
+import { BillingAuthService } from '@/modules/billing/domain/billing-auth.service';
+
+const ISSUER = 'safe-client-gateway';
```
Comment (PooyaRaki) [inline src/modules/billing/domain/billing-auth.service.spec.ts]: Should be a random string.

File: src/modules/billing/domain/billing-auth.service.spec.ts [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+      const service = createService(publicKey);
+
+      const token = BillingAuthService.mint({
+        privateKey,
+        issuer: ISSUER,
+        expiresInDays: 365,
+      });
+
+      expect(service.verify(token)).toMatchObject({
+        roles: ['SERVICE_ACCESS'],
```
Comment (PooyaRaki) [inline src/modules/billing/domain/billing-auth.service.spec.ts]: It's a bit unclear in this test where the `SERVICE_ACCESS` value is coming from and why, I know that's the default. Can we randomly pass a role when we mint a token and make sure that the role is returned when we decode the token?

File: src/modules/billing/domain/billing-auth.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+      // billing.webhook.publicKey intentionally not set.
+
+      expect(
+        () =>
+          new BillingAuthService(
+            new JwtService(jwtClientFactory(), configurationService),
+            configurationService,
+            loggingService,
+          ),
+      ).toThrow();
```
Comment (PooyaRaki) [inline src/modules/billing/domain/billing-auth.service.spec.ts]: Nit: We should check the exact error here.

**Review by PooyaRaki — DISMISSED** [review]

File: src/modules/billing/domain/billing-auth.service.ts (hunk 5 above)
Comment (LucieFaire) [reply src/modules/billing/domain/billing-auth.service.ts]: i moved it to `configuration.ts` directly becuase it wont be applied in schema or validator as they are used strictly as a pass/fail gate.

File: src/modules/billing/routes/guards/billing-webhook-auth.guard.ts (hunk 4 above)
Comment (LucieFaire) [reply src/modules/billing/routes/guards/billing-webhook-auth.guard.ts:20]: From my analysis, the similarity is mostly superficial, and merging would couple two security paths. I would argue against coupling these two together

File: src/modules/billing/domain/billing-auth.service.spec.ts (hunk 9 above)
Comment (LucieFaire) [reply src/modules/billing/domain/billing-auth.service.spec.ts]: The role can't be randomized as it's not a mint input. `buildClaims` specifically sets roles: [SERVICE_ACCESS_ROLE] in order to mint exactly one credential type (a service token) as per contract. Adding a role parameter would contradict that.

Comment (LucieFaire) [reply src/modules/billing/domain/billing-auth.service.spec.ts]: but i changed the test to check for a constant value and not a literal one :)

**Review by moisses89 — DISMISSED** [review]

File: src/modules/billing/routes/guards/billing-webhook-auth.guard.integration.spec.ts [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+import configuration from '@/config/entities/__tests__/configuration';
+import { JWT_ES_ALGORITHM } from '@/datasources/jwt/jwt.constants';
+import { jwtClientFactory } from '@/datasources/jwt/jwt.module';
+import { TestLoggingModule } from '@/logging/__tests__/test.logging.module';
+import { BillingModule } from '@/modules/billing/billing.module';
+import { BillingAuthService } from '@/modules/billing/domain/billing-auth.service';
+import { BillingController } from '@/modules/billing/routes/billing.controller';
+import { BillingWebhookAuthGuard } from '@/modules/billing/routes/guards/billing-webhook-auth.guard';
+
+const ISSUER = 'safe-client-gateway';
```
Comment (PooyaRaki) [inline src/modules/billing/routes/guards/billing-webhook-auth.guard.integration.spec.ts]: Nit: We should use `faker` here as well

Comment (LucieFaire) [reply src/modules/billing/routes/guards/billing-webhook-auth.guard.integration.spec.ts]: [01c4018](https://github.com/safe-global/safe-client-gateway/pull/3217/commits/01c40180571914632cee93438c2392ff7ee20d75)

File: scripts/generate-token.ts [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        'expires-in': { type: 'string' },
+      },
+    });
+    return { sub: values.sub, expiresIn: values['expires-in'] };
+  } catch (error) {
+    fail(error instanceof Error ? error.message : 'Invalid arguments');
+  }
+}
+
+async function main(): Promise<void> {
```
Comment (PooyaRaki) [inline scripts/generate-token.ts]: This method is quite large and very hard to follow, especially with the local private key logic which is handled in an `if` condition. We should extract it to several smaller methods for maintainability.

File: scripts/generate-token.ts [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+        .export({ format: 'pem', type: 'spki' })
+        .toString();
+    } else {
+      // Local PEM signing mode — dev/CI only; production must sign via KMS.
+      if (config.application.isProduction) {
+        fail(
+          'Production requires KMS signing: set BILLING_WEBHOOK_JWT_KMS_KEY_ID. Minting with a local BILLING_WEBHOOK_JWT_PRIVATE_KEY is not allowed in production.',
+        );
+      }
+      const privateKey = process.env.BILLING_WEBHOOK_JWT_PRIVATE_KEY?.replace(
```
Comment (PooyaRaki) [inline scripts/generate-token.ts]: Nit: We could also check this in the env variable validator and ensure it doesn't exist in production.

File: src/datasources/kms/aws-kms-signer.service.spec.ts [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+} from '@aws-sdk/client-kms';
+import { faker } from '@faker-js/faker';
+import { mockClient } from 'aws-sdk-client-mock';
+import { AwsKmsSignerService } from '@/datasources/kms/aws-kms-signer.service';
+
+const kmsMock = mockClient(KMSClient);
+
+describe('AwsKmsSignerService', () => {
+  const keyId = faker.string.uuid();
+  const config = { keyId };
```
Comment (PooyaRaki) [inline src/datasources/kms/aws-kms-signer.service.spec.ts]: It's used only once on line 23, we could remove the `const` and pass the object directly to the constructor.

File: src/datasources/kms/aws-kms-signer.service.ts [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+import {
+  GetPublicKeyCommand,
+  KMSClient,
+  SignCommand,
+} from '@aws-sdk/client-kms';
+import { fromTokenFile } from '@aws-sdk/credential-provider-web-identity';
+import type { IKmsSigner } from '@/datasources/kms/kms-signer.interface';
+
+export interface KmsSignerConfig {
```
Comment (PooyaRaki) [inline src/datasources/kms/aws-kms-signer.service.ts]: Interface should be extracted to another file.

File: src/datasources/kms/aws-kms-signer.service.ts [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+  keyId: string;
+  webIdentityTokenFile?: string;
+}
+
+function resolveCredentials(
```
Comment (PooyaRaki) [inline src/datasources/kms/aws-kms-signer.service.ts]: Helpers should also be extracted to another file.

File: src/datasources/kms/aws-kms-signer.service.ts [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  config: KmsSignerConfig,
+): ReturnType<typeof fromTokenFile> | undefined {
+  // EKS provides pod credentials via IRSA (web identity token). Otherwise fall
+  // back to the default AWS provider chain (env keys).
```
Comment (PooyaRaki) [inline src/datasources/kms/aws-kms-signer.service.ts]: QQ: Should we use dedicated credentials for this instance instead of the default ones?

File: src/modules/billing/domain/billing-auth.service.ts [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+} from '@/logging/logging.interface';
+import {
+  type BillingServiceToken,
+  BillingServiceTokenSchema,
+  SERVICE_ACCESS_PERMISSION_TYPE,
+  SERVICE_ACCESS_ROLE,
+  SERVICE_USER_TYPE,
+} from '@/modules/billing/domain/entities/billing-service-token.entity';
+
+export const DEFAULT_BILLING_SERVICE_TOKEN_SUBJECT = 'billing-service';
```
Comment (PooyaRaki) [inline src/modules/billing/domain/billing-auth.service.ts]: const and `type` should be extracted to a separate file.

File: src/modules/billing/domain/billing-auth.service.ts [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+   *   2. Authorization — confirm it is a service token.
+   */
+  verify(token: string): BillingServiceToken {
+    try {
+      // Layer 1: signature + standard claims.
+      const payload = this.jwtService.verify<BillingServiceToken>(token, {
+        secretOrPrivateKey: this.publicKey,
+        algorithms: [JWT_ES_ALGORITHM],
+        issuer: this.issuer,
+        audience: this.issuer,
```
Comment (PooyaRaki) [inline src/modules/billing/domain/billing-auth.service.ts:72]: Shouldn't the issuer be different from the audience? In this case I expect the issuer to be the billing service and the audience to be `CGW`

**Comment by PooyaRaki** [comment]
@claude review

File: src/datasources/kms/aws-kms-signer.service.ts (hunk 17 above)
Comment (LucieFaire) [reply src/datasources/kms/aws-kms-signer.service.ts]: in practise we should not need static params as everything would be in token file. This is just a fallback, that would fail if we invoke the methods. I think we can leave it as is

File: src/modules/billing/domain/billing-auth.service.ts (hunk 19 above)
Comment (LucieFaire) [reply src/modules/billing/domain/billing-auth.service.ts:72]: CGW issues the token, provides it to billing-service, so that billing-service can use it to auth in CGW. Thus issuer and audience are both CGW. We have the same pattern for auth0

**Comment by LucieFaire** [comment]
@claude review again

**Comment by LucieFaire** [comment]
Claude review:
Addressed (2,5,6,8), (1,3,4,7) is irrelevant, added extra tests (9-11) and resolved the nits.

**Comment by PooyaRaki** [comment]
@claude review
As @LucieFaire said:

> Addressed (2,5,6,8), (1,3,4,7) is irrelevant, added extra tests (9-11) and resolved the nits.

Check all the previous remarks from yourself and other reviewers.

**Review by PooyaRaki — APPROVED** [review]
